### PR TITLE
Pre-filter out problematic buffers upon close

### PR DESCRIPTION
**Pull request by jordwalke**
**Bug:**
```
:e ~/existing.re
:vsplit ~/anyOtherFile (Then modify this new vert split unnamed buffer)
:WintabsClose  (It will error)
```
I think before doing anything in WinClose, we should first get the file to  "a non-problematic state". That means revert it if it's not opened in any other tab. This way we don't have to deal with the endless complexities that vim throws at us with listed/unlisted buffers - where we don't find out about errors before it's too late.

This might allow you to remove other parts of the code that acts defensively.

### CODE REVIEW

**Comment by zefei** [comment]
This looks like a better approach than my reliance on vim's `nohidden+confirm` behavior. I'll try this patch and probably merge it manually.

**Comment by jordwalke** [comment]
I'm using it and it's pretty reliable so far. We need more testing though. The downside was that I had to come up with something that works in all cases like unnamed buffers, buffers where the undo stack is past/before the last save point. Vim really should have a command "return to the last save point for this buffer".

There's one *minor* downside to my current approach. If you made changes without saving, that would ideally leave "redo" commits when you reopen the file next time, if you use Vim's persistent (awesome) undo like I do. With this approach, the `:e!` command will remove those if you choose not to save. It's pretty minor - it's as if you had done `:e!` yourself before closing the file.

But there's no other way it seems. Vim doesn't have a way to "return to last saved position"  - any combination of `:earlier 1f/0f` / `:later 1f/0f` had some problem.

File: autoload/wintabs.vim [hunk 1]
```diff
@@ -42,10 +42,41 @@ endfunction
...
+  " unlisted buffers.
+  let occurrence = s:count_occurrence(buffer)
+  if getbufvar(buffer, '&modified') && occurrence < 2
+    let choice = confirm(msg . "\n", "&Save\n&Cancel\n&Don't Save", 2)
+    if choice == 3
+      " Don't save: Reset the file then call WintabsClose
+      " earlier 1f doesn't always work (like if you already did undo a few
+      " times without saving), but it does work on brand new files that have
+      " no name!
+      if name == ''
```
Comment (jordwalke) [inline autoload/wintabs.vim]: Is there another/better way to detect "unnamed buffer"  - I meant to be checking the case where you do `:enew` and make changes, but don't save before closing.

File: autoload/wintabs.vim [hunk 2]
```diff
@@ -42,10 +42,41 @@ endfunction
...
 " close current tab
 function! wintabs#close()
-  call wintabs#refresh_buflist(0)
 
-  let size = len(w:wintabs_buflist)
   let buffer = bufnr('%')
+  let name = expand('%:t')
+  let msg = 'Do you want to save the changes you made to document '
+  let msg = msg . (name == '' ? "\"Untitled\"\n" : (name . "\n"))
+  let msg = msg . "Your changes will be lost if you don't save them."
```
Comment (jordwalke) [inline autoload/wintabs.vim]: Tested this message in MacVim and terminal Vim (8). The message matches the GUI that MacVim shows.

**Comment by jordwalke** [comment]
After this, I think we should add a similar feature for when you do `:WintabsCloseWindow`. When you do something like closing multiple tabs/windows in MacVim, it checks if there exists *any* buffer in all the tabs that are about to be closed and asks you what to do. If there's just one with unsaved changes, it has the same experience as this PR. If there's multiple - it gives you several options - we can recreate this:

screen shot 2018-03-02 at 9 23 21 pm

It cycles one by one through all the unsaved docs, asking you what to do with each one but also letting you select "Discard All" to stop cycling and abandon all changes.

**Comment by zefei** [comment]
@jordwalke I implemented something similar in the recent commits - try to save buffer at the beginning of wintabs#close instead of doing it using nohidden+confirm. But instead of using a custom save dialog, I'm relying on `confirm edit` since this dialog is i18n'ed.

I'm closing this PR, but please let me know if there are bugs in my implementation.

**Pull request closed without merging**